docs/design/alto/tpc-engine.md

```diff
@@ -0,0 +1,146 @@
+processed on the eventloop. Interaction with the reactor is thread-safe.
+
+A reactor is configured/created using the ReactorBuilder. There is a
+ReactorBuilder for NIO, and there will be a ReactorBuilder for io_uring
+because each reactor can have its own implementation-specific
+properties.
+
+The eventloop thread can be pinned to a core or a set of cores, but that
+isn’t done by default.
+
+## Eventloop
+
+The eventloop is a loop which processes events. This can be I/O events,
+but also ‘tasks’. It is very similar to the eventloop in Javascript. The
+eventloop will wait for async I/O and tasks ‘events’ and will call the
+appropriate handler for that event and just run the tasks (like an
+executor).
+
+The eventloop should only be accessed from the owning event-loop-thread
+because the Eventloop isn’t thread-safe.
+
+The eventloop also has the ability for task scheduling:
+
+1. Schedule a task to run at some point in time
+2. Schedule a task to be run with a fixed delay or fixed frequency
+
+These tasks are run on the event loop-thread itself (since we do not
+want other threads).
+
+The eventloop is specific to the async I/O library used. In the 5.3
+release, we ship the NioEventloop (and NioReactor) which can only do
+networking. On Linux, the NioEventloop is EPoll based and the
+non-blocking I/O functionality is provided by the java.nio.Selector.
+
+In the 5.4 release, we will also add the IOUringEventloop (and
+IOUringReactor) that can do both networking and storage and will rely on
+the io_uring for non-blocking I/O. Probably the IOUringReactor will be
+part of the Hazelcast enterprise. The IOUringEventloop is specific to
+Linux and requires linux 5.7+.
+
+The eventloop can spin instead of block on the I/O multiplexer, but that
+isn’t done by default.
+
+## Promise
+
+You never want to block the Eventloop because it will stall anything
+running on top of this eventloop. So anything that otherwise would block
+needs to be done asynchronously, and you need to have some kind of
+mechanism to continue once some async task completes. That is the task
+of the promise. The Promise is the TPC version of the CompletableFuture.
+
+The Promise can only be used within a single eventloop and isn’t
+threadsafe. Promises support pooling to prevent create litter on the
+fast path.
+
+## TPCEngine
+
+The TPCEngine is effectively nothing else than an array of reactors. It
+is mostly a matter of convenience, it doesn't provide much more
+functionality apart from creation, starting, and stopping.
+
+Reactors can be used directly without the need for a TPCEngine. This
+makes testing a lot less cumbersome.
+
+The functionality of the tpc-engine doesn't know anything about
+functionality running on top like Alto or Hazelcast classic.
+
+The TpcEngine is created using the Configuration object. Probably this
+should be renamed to TpcEngineBuilder since most of the other objects
+are created using Builders.
```

Comment:
Wouldn't it be better to do it directly within this PR?

Reply:
https://github.com/hazelcast/hazelcast/pull/23853, let me also update but note that I just copied the referenced link in the tdd.

Reply:
@kwart no. There is already a different PR with that change. It would have been better of the TDD of the tpc-engine would have been added into its own PR. 

Now a lot of changes, which are completely independent of each, other get lumped into a single PR.